### ISSUE TITLE
Differences in class instantiation release/debug VS 2012 vs. VS 2015/2017 

### ISSUE THREAD

**Issue by varocarbas** [dotnet/roslyn#19346]
**Version Used**: 
Visual Studio 2012, 2015, 2017 (exact versions are being displayed in the video).

**Steps to Reproduce**:
I have recorded a [video clearly showing the problem](https://www.youtube.com/watch?v=d__lXHqmPFE). Note that this refers to the library DateParser, whose code is public and can be found [here](https://github.com/varocarbas/FlexibleParser/tree/master/all_code/DateParser) (it shares the same solution with other libraries; you can just download everything). 

In a nutshell, the problem is that some public variables of a class, which are used as a way to call a function where various required actions are performed, aren't being hit at class instantiation under specific conditions (newer VS and release modes).

**Expected Behavior**:
Debug /release modes behaving identically (either error or no error). All the VS versions having the same (even faulty) behaviour.

**Actual Behavior**:
VS 2012 behaves fine (both release/debug don't trigger any error). VS 2015/2017 work fine in debug mode, but trigger an error in release mode.

CLARIFICATION: although I do consider that this is a faulty behaviour for various reasons (e.g., differences debug/release and between VS versions), I am not interested in actively participating in a discussion about all the associated issues like the reasons for it to be there, alternative approaches (for anyone interested: I will modify the code such that these conditions will be plainly avoided) or the numerous difficulties associated with performing an eventual modification. The only motivation of this issue is to let you know about what I think that might be of interest to you; am also providing more than enough information, that's why I don't think that my participation is strictly required.

**Comment by jaredpar** [comment]
> VS 2015/2017 work fine in debug mode, but trigger an error in release mode.

Can you elaborate on what error is triggered? The video flashes a bunch and makes it really hard to watch. Would help to see a solid code sample + what behavior differs.

**Comment by varocarbas** [comment]
@jaredpar 

> The video flashes a bunch
It is fine for me. Have you tried to increase its quality?

> solid code sample + what behavior differs.
There are step-by-step instructions in the video description explaining how to get the version of the code provoking exactly the shown behaviour. Anyway, I can try to write a much smaller code provoking that situation. 
In any case, note that the referred library isn't precisely simple (even just the part involved in that misbehaviour) and I am not completely sure about the exact conditions triggering the problem (that's why I add a reference to the whole code allowing everyone to replicate the exact same conditions). In principle, I should be able to write a more descriptive code quite quickly (in case it isn't clear yet: it is about changing the order/skipping the analysis of some public variables when instantiating the class), at some point today. Otherwise, it will have to wait until next week when I have more free time.

**Comment by varocarbas** [comment]
@jaredpar

It was quicker than what I thought! Here you have your clearly-descriptive simplistic code, written in C# and classic console.

```C#
using System;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            Class.PublicMethod();
        }
    }

    public class Class
    {
        public static string PublicVariable = PublicVariableHit();

        private static string PublicVariableHit()
        {
            Console.WriteLine("Public variable hit at class instatiation.");

            return "";
        }

        public static void PublicMethod()
        {
            Console.WriteLine("Public class method started.");
            Console.ReadLine();
        }
    }
}
```

If you execute this code in the referred VS (in its default configuration, the versions displayed in the video. Just in case that aspect might also be relevant, note that I am running Windows 7 64-bit), you should get the following:
- VS 2012: both messages are displayed in debug and release modes.
- VS 2015 & 2017: both messages are displayed in debug mode, but only "Public class method started." in release mode.

**Comment by sharwell** [comment]
@varocarbas This behavior is covered by the language on [Static field initialization](https://github.com/dotnet/csharplang/blob/master/spec/classes.md#static-field-initialization) in the C# Language Specification:

> The static field variable initializers of a class correspond to a sequence of assignments that are executed in the textual order in which they appear in the class declaration. If a static constructor (Static constructors) exists in the class, execution of the static field initializers occurs immediately prior to executing that static constructor. Otherwise, the static field initializers are executed at an implementation-dependent time prior to the first use of a static field of that class.

Since `PublicMethod()` does not access `PublicVariable` (and nor does any code invoked from `Main()`), it is not required that `PublicVariableHit()` be called unless you add a static constructor to `Class`:

```csharp
static Class()
{
  // Require all fields be initialized before this class is used
}
```

**Comment by varocarbas** [comment]
@sharwell 
I don't want to be rude, but I think that my initial post is quite clear regarding my intention, expectations and the kind of participation which I was interested in having here.

Just for the record, different behaviours release/debug (and additionally between different VS versions!) is a no-brainer to me: certainly a bug. Choose the behaviour you want and make sure that it works exactly in that way in each single scenario.

But you might even ignore the previous paragraph and focus on the aforementioned clear intention regarding my zero interest in getting involved in any kind of discussion here regarding what should/shouldn't be. I think that I have been quite nice by performing an additional work when requested (one which, honestly,  I don't even think that is too compatible with my original request as far as the video + original code shouldn't have been more than enough). 

If you think that this reaction a bit too excessive, you should take a look at quite a few of my previous interactions here + my final decisions of not coming back unless strictly required (e.g., sharing information which you might want to know).

**Comment by sharwell** [comment]
@varocarbas That's fine. Your example was very helpful in narrowing down why the behavior difference might be appearing. Since some things move slow (changes to the compiler) while other things move fast (many external library projects), I was commenting above to cover two things:

1. Why the behavior change might be occurring (i.e. the language specifically states that under this condition it is allowed to have different behaviors, and may even differ in behavior between two runs of the same application)
2. A step someone could take to make the behavior deterministic (i.e. the behavior would not change between Debug/Release or between different releases of the compiler)

To be more specific, a minimum change that would be needed in the library code would be changing this code to something like this:

```csharp
static TimeZoneMilitary()
{
  TimeZonesInternal.StartTimezones();
}
```

:memo: I only investigated a minimum portion of the code based on the fragment seen in the video. If you are interested in a more detailed review to find cases where bugs might be lurking due to an unintentional reliance on [unspecified behavior (§B.3)](https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-334.pdf#page=496), I'm happy to help out if you ask.

:bulb: I noticed that you use a coding style where classes are often split up. This is fine (I make no claims that my comfort zone needs to be yours), but I wanted to point out another reason why static constructors would be of interest to you: the specification doesn't require a particular order for field initializers found in different files, such as [`TimeZoneMilitary.Populated`](https://github.com/varocarbas/FlexibleParser/blob/c165ab9c45dca4bdb06386ec9a0eda134a675f17/all_code/DateParser/Source/TimeZones/Types/Military/TimeZones_Types_Military_Constructors.cs#L7) and [`TimeZoneMilitary.MainType`](https://github.com/varocarbas/FlexibleParser/blob/c165ab9c45dca4bdb06386ec9a0eda134a675f17/all_code/DateParser/Source/TimeZones/Types/Military/TimeZones_Types_Military_Methods.cs#L7). However, if you initialize these fields in a static constructor, then you can guarantee the order of initialization will never change. If you ever aren't happy with the reorganization that causes, you can selectively change certain fields to be expression-bodied properties. For example, `MainType` performs essentially the same using the following code except for the most performance-critical application where it is extremely frequently referenced:

```csharp
private static Type MainType => typeof(TimeZoneMilitary);
```

**Comment by varocarbas** [comment]
@sharwell
I think that you re-read my first post because you don't seem to have clear some ideas which are explained there.

> To be more specific, a minimum change that would be needed in the library code would be changing this code to something like this:
 
All what follows this statement is incompatible with the original "for anyone interested: I will modify the code such that these conditions will be plainly avoided" (FYI, I corrected the original code of DateParser a short while after having posted this issue. That's why you need instructions to download a code with this error, the last version doesn't have it)

> I'm happy to help out if you ask.

I am an experienced programmer (in C# and in quite a few other languages; mostly experienced in developing efficient algorithms completely from scratch). I don't ask, I answer :). Honestly, I am eminently experienced in making things actually work by being only reasonably-concerned about the code structure/beauty; DateParser (and the previous FlexibleParser parts) are my first relevant enough open-source code where I am paying more attention than usual to those aspects (the experience has been quite positive, as far as my I have somehow evolved the way in which I face a code by giving much more thought to its structure/scalability). The reason why I say that I don't belong here is because the discussions here aren't precisely eminently programming-focused, understood as aiming for the objectively best solutions. There are also lots of misunderstandings and even not-too-clear motivations regarding some actions. Again, I don't want to offend anyone, not even to blame others of what might have been a normal reaction to my behaviour; but plainly sharing the current reality: I prefer to not get involved in discussions here. I shared information which you might find interesting, use it as you wish or not.

> I noticed that you use a coding style where classes are often split up. This is fine (I make no claims that my comfort zone needs to be yours), but I wanted to point out another reason why static constructors would be of interest to you: the specification doesn't require a particular order for field initializers found in different files, such as TimeZoneMilitary.Populated and TimeZoneMilitary.MainType. However, if you initialize these fields in a static constructor, then you can guarantee the order of initialization will never change. If you ever aren't happy with the reorganization that causes, you can selectively change certain fields to be expression-bodied properties. For example, MainType performs essentially the same using the following code except for the most performance-critical application where it is extremely frequently referenced:

You are referring to the old code. Just FYI, I came up with this curious way (= never did it before) to populate the original information on the go, almost for fun. The code kept growing and this part remained, until I realised about that bug by pure chance (because I performed all the development/testing in VS 2012). Then, I looked at that part and realised that wasn't even a good idea at that point on account of the new structure of classes (I developed that bit at the very beginning when the number of independent classes/constructors was much more relevant than what finally became). So, actually speaking, that curious approach which allowed me to find this bug and you are trying to correct isn't even a normal output of my usual coding approach, it happened to be there and I removed it right away as soon as it was proven problematic.

**Comment by sharwell** [comment]
@varocarbas Thank you for the feedback. I filed #20268 with one idea for a way we can be proactive about helping other developers avoid similar situations. I have a feeling it may end up moved to [dotnet/roslyn-analyzers](https://github.com/dotnet/roslyn-analyzers) but it should provide a solid starting point.

**Comment by varocarbas** [comment]
@sharwell Thanks for being so understanding.

**Comment by VSadov** [comment]
@sharwell is right about:

> the static field initializers are executed at an implementation-dependent time prior to the first use of a static field of that class.

A type without a formal static constructor is emitted with a "beforefieldinit" flag which permits the runtime to be lazy or eager about the initialization as long as the field is initialized before the first use. Runtimes do take advantage of this for the purpose of improving performance. 
(a good article touching that subject here: http://csharpindepth.com/Articles/General/Beforefieldinit.aspx)

What you are observing is largerly driven not by the differences in the compiled code, but by differences in the runtime environment.

In fact - with the repro sample - if you change the target framework of your 2017 project to anything before `.NET Framework 4` you are likely to see the a behavior compatible with "2012". 
With the exception of the target framework attribute, compiler will produce the same code, but runtime will observe the attribute and try having more compatible behavior.

**Comment by varocarbas** [comment]
@VSadov

> A type without a formal static constructor is emitted with a "beforefieldinit" flag which permits the runtime to be lazy or eager about the initialization as long as the field is initialized before the first use. Runtimes do take advantage of this for the purpose of improving performance.
> (a good article touching that subject here: http://csharpindepth.com/Articles/General/Beforefieldinit.aspx)

As said to @sharwell, thanks for the information but I am not too concerned about this kind of scenarios as far as my code virtually never deals with them. The referred library (and even the other parts of FlexibleParser) was almost an experiment which happened to reach a quite unusual outcome. In any case, I think that this reality is perfectly compatible with the true reason for my ticket (+ my consideration of this situation being a bug): different behaviours release/debug and between different VS versions. Do whatever you wish, but be consistent and compatible. 

> What you are observing is largerly driven not by the differences in the compiled code, but by differences in the runtime environment.
> 
> In fact - with the repro sample - if you change the target framework of your 2017 project to anything before .NET Framework 4 you are likely to see a behavior compatible with "2012".
> With the exception of the target framework attribute, compiler will produce the same code, but runtime will observe the attribute and try having more compatible behavior.

I didn't analyse the specific conditions in too much detail, but the .NET Framework version sounds as a perfect candidate for the problem. The lack of compatibility being due to different .NET Frameworks or VSs sounds equally bug-label worthy to me.

I have no interest in extending this conversation further or in convincing anyone about anything. I plainly shared some information about which I thought that you might want to know.

**Comment by sharwell** [comment]
For anyone else interested, I discussed this with some people internally to try and better understand how a change in behavior appeared even when we try to avoid observable changes to the degree reasonably possible. In this case, it turns out everyone's hands were essentially tied because:

1. Since the original specification allowed a degree of variance in runtime behavior, no one ever documented the precise behavior of the implementation under all edge cases. (This is unspecified behavior, as opposed to implementation-defined behavior which would have required this analysis.)
2. The behavior changed in historical releases, sometimes subtly, such that even if one release was fully-understood it would be fundamentally impossible to maintain consistent behavior across all releases.

Since the behavior deviation was allowed and had already deviated in the past, the policy moving forward was to continue allowing minor deviations in this space to achieve proven wins for performance at runtime. Changing the policy at this point would both hinder the ability to realize performance gains *and* not actually achieve perfect behavior consistency.
